From 774b701ae1bab296ec58beaf99fdebcb8723e3c8 Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Thu, 13 Jun 2013 13:33:53 +0800
Subject: [PATCH 18/29] perf-top: Show the percentage of successfull PEBS-fixups

commit 1676b8a077c352085d52578fb4f29350b58b6e74 upstream

Use the PERF_RECORD_MISC_EXACT information to measure the success rate of
the PEBS fix-up.

Signed-off-by: Peter Zijlstra <a.p.zijlstra@chello.nl>
Cc: Arnaldo Carvalho de Melo <acme@infradead.org>
Cc: paulus@samba.org
Cc: eranian@google.com
Cc: robert.richter@amd.com
Cc: fweisbec@gmail.com
LKML-Reference: <20100304140100.694233760@chello.nl>
Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Xufeng Zhang <xufeng.zhang@windriver.com>
---
 tools/perf/builtin-top.c |   21 +++++++++++++++------
 1 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/tools/perf/builtin-top.c b/tools/perf/builtin-top.c
index 0f6d18f..3acc3ac 100644
--- a/tools/perf/builtin-top.c
+++ b/tools/perf/builtin-top.c
@@ -418,6 +418,7 @@ static double sym_weight(const struct sym_entry *sym)
 
 static long			samples;
 static long			kernel_samples, us_samples;
+static long			exact_samples;
 static long			guest_us_samples, guest_kernel_samples;
 static const char		CONSOLE_CLEAR[] = "[H[2J";
 
@@ -462,6 +463,7 @@ static void print_sym_table(void)
 	float us_samples_per_sec = (us_samples)/delay_secs;
 	float guest_kernel_samples_per_sec = (guest_kernel_samples)/delay_secs;
 	float guest_us_samples_per_sec = (guest_us_samples)/delay_secs;
+	float esamples_percent = (100.0*exact_samples)/samples;
 	float sum_ksamples = 0.0;
 	struct sym_entry *syme, *n;
 	struct rb_root tmp = RB_ROOT;
@@ -469,7 +471,7 @@ static void print_sym_table(void)
 	int sym_width = 0, dso_width = 0, dso_short_width = 0;
 	const int win_width = winsize.ws_col - 1;
 
-	samples = us_samples = kernel_samples = 0;
+	samples = us_samples = kernel_samples = exact_samples = 0;
 	guest_kernel_samples = guest_us_samples = 0;
 
 	/* Sort the active symbols */
@@ -502,13 +504,16 @@ static void print_sym_table(void)
 
 	printf("%-*.*s\n", win_width, win_width, graph_dotted_line);
 	if (!perf_guest) {
-		printf( "   PerfTop:%8.0f irqs/sec  kernel:%4.1f%% [",
+		printf( "   PerfTop:%8.0f irqs/sec  kernel:%4.1f%%"
+			"  exact: %4.1f%% [",
 			samples_per_sec,
 			100.0 - (100.0 * ((samples_per_sec - ksamples_per_sec) /
-					samples_per_sec)));
+					samples_per_sec)),
+			esamples_percent);
 	} else {
 		printf("   PerfTop:%8.0f irqs/sec  kernel:%4.1f%% us:%4.1f%%"
-				" guest kernel:%4.1f%% guest us:%4.1f%% [",
+			" guest kernel:%4.1f%% guest us:%4.1f%%"
+			" exact: %4.1f%% [",
 				samples_per_sec,
 				100.0 - (100.0 * ((samples_per_sec-ksamples_per_sec) /
 						samples_per_sec)),
@@ -518,8 +523,9 @@ static void print_sym_table(void)
 							guest_kernel_samples_per_sec) /
 						samples_per_sec)),
 				100.0 - (100.0 * ((samples_per_sec -
-							guest_us_samples_per_sec) /
-						samples_per_sec)));
+						guest_us_samples_per_sec) /
+						samples_per_sec)),
+				esamples_percent);
 	}
 
 	if (nr_counters == 1 || !display_weighted) {
@@ -1011,6 +1017,9 @@ static void event__process_sample(const event_t *self,
 		return;
 	}
 
+	if (self->header.misc & PERF_RECORD_MISC_EXACT)
+		exact_samples++;
+
 	if (event__preprocess_sample(self, session, &al, symbol_filter) < 0 ||
 	    al.filtered)
 		return;
-- 
1.7.0

